wagtail/admin/viewsets/base.py

```diff
@@ -28,6 +28,49 @@ def __init__(self, name=None, **kwargs):
         for key, value in kwargs.items():
             self.__dict__[key] = value
 
+    def get_common_view_kwargs(self, **kwargs):
+        """
+        Returns a dictionary of keyword arguments to be passed to all views within this viewset.
+        """
+        return kwargs
+
+    def construct_view(self, view_class, **kwargs):
+        """
+        Wrapper for view_class.as_view() which passes the kwargs returned from get_common_view_kwargs
+        in addition to any kwargs passed to this method. Items from get_common_view_kwargs will be
+        filtered to only include those that are valid for the given view_class.
+        """
+        filtered_kwargs = {
+            key: value
+            for key, value in self.get_common_view_kwargs().items()
+            if hasattr(view_class, key)
+        }
+        filtered_kwargs.update(kwargs)
+        return view_class.as_view(**filtered_kwargs)
+
+    def inject_view_methods(self, view_class, method_names):
+        """
+        Check for the presence of any of the named methods on this viewset. If any are found,
+        create a subclass of view_class that overrides those methods to call the implementation
+        on this viewset instead. Otherwise, return view_class unmodified.
+        """
+        viewset = self
+        overrides = {}
+        for method_name in method_names:
+            viewset_method = getattr(viewset, method_name, None)
+            if viewset_method:
+
+                def view_method(self, *args, **kwargs):
+                    return viewset_method(*args, **kwargs)
```

Comment:
I'm wondering whether we should allow the viewset method overrides to have access to the view object? e.g. something like

```suggestion
                def view_method(self, *args, **kwargs):
                    return viewset_method(*args, view=self, **kwargs)
```

Then we'll need to enforce that view methods that are overridden on the viewset should have an additional `view` parameter, but it has the added benefit of having access to the view's instance attributes e.g. `request`.

Reply:
See my previous response at https://github.com/wagtail/rfcs/pull/87#discussion_r1311391102 - I'm not convinced that making the view available here is a good idea, especially as the whole point of putting methods on the viewset is that we don't want the logic to be coupled to one particular view.

I've been going back and forth on this decision a bit - in my short time using this feature I've already found myself wanting to filter the object list by URL parameters, which would be a natural use-case for this. On the other hand, I've _also_ found myself wanting to use `super` here:

```python
    def get_object_list(self):
        return super().get_object_list().filter(is_active=True)
```

...which definitely won't work without a lot more trickery. For that reason, I think we're better off embracing the fact that methods on the viewset aren't going to be as capable as methods defined on the views themselves, and that if you want to do anything that relies on tapping into the state of the views, you should override the view instead.

Reply:
Fair enough! :D